core_backend/app/contents/routers.py

```diff
@@ -158,6 +165,209 @@ async def retrieve_content_by_id(
     return _convert_record_to_schema(record)
 
 
+@router.post("/csv-upload", response_model=List[ContentRetrieve])
+async def bulk_upload_contents(
+    file: UploadFile,
+    user_db: Annotated[UserDB, Depends(get_current_user)],
+    asession: AsyncSession = Depends(get_async_session),
+) -> List[ContentRetrieve] | None:
+    """
+    Upload, check, and ingest contents in bulk from a CSV file.
+
+    Note: If there are any issues with the CSV, the endpoint will return a 400 error
+    with the list of issues under detail in the response body.
+    """
+
+    # TODO: deal with tags!
+
+    # Ensure the file is a CSV
+    if file.filename is None or not file.filename.endswith(".csv"):
+        error_list_model = CustomErrorList(
+            errors=[
+                CustomError(
+                    type="invalid_format",
+                    description="Please upload a CSV file.",
+                )
+            ]
+        )
+        raise HTTPException(status_code=400, detail=error_list_model.dict())
+
+    df = _load_csv(file)
+    await _csv_checks(df=df, user_id=user_db.user_id, asession=asession)
+
+    # Add each row to the content database
+    content_list = []
+    for _, row in df.iterrows():
+        content = ContentCreate(
+            content_title=row["content_title"],
+            content_text=row["content_text"],
+            content_language=row.get("content_language", "ENGLISH"),
+            content_tags=[],
+            content_metadata=row.get("content_metadata", {}),
+        )
+        content_db = await save_content_to_db(
+            user_id=user_db.user_id, content=content, asession=asession
+        )
+        content_retrieve = _convert_record_to_schema(content_db)
+        content_list.append(content_retrieve)
+
+    return content_list
+
+
+def _load_csv(file: UploadFile) -> pd.DataFrame:
+    """
+    Load the CSV file into a pandas DataFrame
+    """
+
+    try:
+        df = pd.read_csv(file.file, dtype=str)
+    except EmptyDataError as e:
+        error_list_model = CustomErrorList(
+            errors=[
```

Comment:
I think it would be neater if you return a bool value and a string specifying the details of the error instead of instanciating a new HTTPException for each error.
Something list this
```
if "content_title" not in cols or "content_text" not in cols:
        return False, "File must have 'content_title' and 'content_text' columns."
```

Reply:
I did it this way so we can get the benefits of Pydantic + keep consistency if we add more checks in the future. Also, these errors are always returned as a list so that we can display them altogether on the front-end if we collect multiple (see _csv_checks). Not 100% the way I've done it is the best though so open to changing - @sidravi1 what do you think?